### PR TITLE
feat(controller): Inline templates. Closes #5105

### LINKED ISSUE #5105: Inlined/anonymous templates
Currently, `DAGTask` and `WorkflowStep` reference templates. What if they were templates themselves? You could inline templates, which I think would be more user friendly, especially making `DataTemplate` easier to use with other templates and reducing the need for more powerful templating features (e.g. Sprig templating or `WithExpression`.

### PR DESCRIPTION
**Pull request by alexec**
image

### CODE REVIEW

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 1]
```diff
@@ -1108,6 +1108,8 @@ type Outputs struct {
 
 // WorkflowStep is a reference to a template to execute in a series of step
 type WorkflowStep struct {
+	Inline *Template `json:"inline,omitempty" protobuf:"bytes,12,opt,name=inline"`
```
Comment (dinever) [inline pkg/apis/workflow/v1alpha1/workflow_types.go]: This is great! Just want to point out that this field is missing the comments.

Screen Shot 2021-04-23 at 5 12 12 PM

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 2]
```diff
@@ -2327,6 +2337,8 @@ type DAGTemplate struct {
 
 // DAGTask represents a node in the graph during DAG execution
 type DAGTask struct {
+	Inline *Template `json:"inline,omitempty" protobuf:"bytes,13,opt,name=inline"`
```
Comment (dinever) [inline pkg/apis/workflow/v1alpha1/workflow_types.go]: This too

File: pkg/apis/workflow/v1alpha1/workflow_types.go (hunk 1 above)
Comment (dinever) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: already fixed 😂  please ignore

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 3]
```diff
@@ -1800,6 +1807,10 @@ func (n NodeStatus) GetTemplateScope() (ResourceScope, string) {
 
 var _ TemplateReferenceHolder = &NodeStatus{}
 
+func (ns *NodeStatus) GetTemplate() *Template {
+	return nil
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:1811]: this could be a problem

**Comment by alexec** [comment]
@jessesuen could I please request your review on this? This is for v3.2, we have not released v3.1, so no hurry.

**Comment by sarabala1979** [comment]
@alexec Quick question about this feature
1. Will it support input/output artifacts and parameters?
2. Can we define the inline in workflowtemplate and refer to it?
3. Inline is referring to template struct.  Will it allow to define the DAG/Step as inline?

**Comment by alexec** [comment]
> * Will it support input/output artifacts and parameters?

Yes.

> * Can we define the inline in workflowtemplate and refer to it?

No. It is not possible to reference inlined templates, by design.

> * Inline is referring to template struct.  Will it allow to define the DAG/Step as inline?

Yes/no. You can inline a DAG within a DAG, or steps within steps etc, but there is the limitation is a single inline.

**Comment by sarabala1979** [comment]
> > * Will it support input/output artifacts and parameters?
> 
> Yes.
> 
> > * Can we define the inline in workflowtemplate and refer to it?
> 
> No. It is not possible to reference inlined templates, by design.

Can we add validation for this?

> 
> > * Inline is referring to template struct.  Will it allow to define the DAG/Step as inline?
> 
> Yes/no. You can inline a DAG within a DAG, or steps within steps etc, but there is the limitation is a single inline.

Can we restrict with  leaf template types like 'container`, `script`, `suspend`, and `resource`?

**Comment by alexec** [comment]
> Can we add validation for this?

I don't think we need to do this for MVP. But later on, maybe.

> Can we restrict with leaf template types like 'container, script, suspend, and resource`?

Again, I don't think we need this for MVP.

**Review by sarabala1979 — APPROVED** [review]

**Pull request merged**